Documentation for Stored Layer Configuration (#243)
sirensolutions/enhanced_tilemap: docs/modules/siren-investigate/pages/enhanced_tilemap.adoc

  * *Popup Content* - Create a comma separated list of fields (in the properties object of your GeoJSON features) to appear on popup tooltip. Note, this is case-sensitive. For example `City_name,Pop_est` would add the city name and estimated population fields to each feature added to the map

==== Stored Layer configurations

This is where configurations for Stored Layers are defined. See <<Loading Stored Layers into Elasticsearch>> for more details on ingesting Stored Layer Data. The following parameters can be configured: 

* *Spatial path* - A path that the configuration usesshould correspond to a layer

NOTE: One configuration is allowed to omit a spatial_path attribute. This is the default configuration and values will be taken from it if they are not found within the <<Cascading Configuration Process>>

* *Icon* - Favicon to display for points on map and layer control
* *Color* - Color to display on map and layer control
* *Size* - Size to display points on map
* *Popup Fields* - The content of the tooltip that appears when mousing over features

> * *Min Visible Zoom* - The minimum map zoom a layer will be visible on the map

Author: ```suggestion
* *Min Visible Zoom* - The zoom level of the map must at least this configuration in order for it to be visible on the map
```

> * *Min Visible Zoom* - The minimum map zoom a layer will be visible on the map

Reviewer: Do you mean, "The zoom level must be set to at least this value for it to be visible on the map." But, what is "it"?

> * *Min Visible Zoom* - The minimum map zoom a layer will be visible on the map

Author: I updated the suggestion, yes that is what I mean, thanks :) - 
![MinZoomInAction](https://user-images.githubusercontent.com/36197976/82378916-eed6b900-9a1d-11ea-82a0-b14550a4e760.gif)

> * *Min Visible Zoom* - The minimum map zoom a layer will be visible on the map

Reviewer: Ok, makes sense now. I think there might be room for a bit more detail about how zoom layers work, for example, what would a minzoom value of 1 represent vs. a minzoom value of 5? If you have time to add that. Otherwise, the update looks good to me.

> * *Max Visible Zoom* - The maximum map zoom a layer will be visible on the map

Author: ```suggestion
* *Max Visible Zoom* - The zoom level of the map must be at most this configuration in order for it to be visible on the map
```

===== Parameter and Field Configurations
It is possible to configure some parameters either using the parameter or field approaches. The parameter approach is where the exact parameter to use on the map is outlined. It is also possible to configure a field that exists within the documents of each layer. Examples of both approaches are outlined in the table below. Parameters are configured by a string, while a single element array is used for the field type approach. 

.Parameter configuration examples
|===
|Parameter |Parameter examples |Field example 

|spatial_path|"World Countries"|N/A

|icon|"fas fa-arrow-alt-circle-down"|["properties.icontouse"]

|color|"#7CBFFA", "blue"|["colortouse"]

|size|"xs", "s", "m", "l", "xl",|["properties.size"]

|popupFields|"is constant for all features"|["fields", "mustbein", "afeatures","propertiesobject"]

|minZoom|0|N/A

|maxZoom|18|N/A
|===

NOTE: When a new Enhanced Coordinate Map visualzation is created, the Stored Layer Configuration will be popuplated with a default object: image:default_default_stored_layers_config.png[Default default stored layers configuration]

> ===== Feature level configurations

Reviewer: ```suggestion
===== Feature-level configurations
```

> If either icon, size, color or popupField are specified in a document's properties object, this field will take precedence. This may be useful for assigning different icons witin the same layer

Reviewer: ```suggestion
If any of the `icon`, `size`, `color`, or `popupField` parameters are specified in a document's `properties` object, this field will take precedence. This can be useful if you want to assign different icons within the same layer.
```

The console output below is a successful file load (with --debug) to Elasticsearch
image:load_map_reference_console_output.png[Console output]

> === Loading Stored Layers from Elasticsearch onto Enhanced Coordinate Map

Reviewer: ```suggestion
=== Loading stored layers from Elasticsearch into the Enhanced Coordinate Map
```

Reviewer: I think 'into' is the right word.. but I am not even confident at this point X(

> In this section we will describe how to load Stored Layers onto the map. 

Reviewer: ```suggestion
In this section, you can learn how to load stored layers into the map. 
```

> Clicking the Add button will add the selected layers to the map, while the Add and Enable button will add the layers to the map and also make them visible. The modal can be cancelled by either clicking the cancel button or the x in the top right corner.

Reviewer: ```suggestion
To add the selected layers to the map but keep them hidden for now, click *Add*.  To add the layers to the map and also make them visible, click *Add and Enable*. 
```

> A number of Stored Layer parameters can be configured from the UI <<Stored Layer Configurations, when in Visualization edit mode>>

Reviewer: ```suggestion
You can configure the stored layers when your visualization is in edit mode. For more information, see <<Stored Layer Configurations, when in Visualization edit mode>>.
```

> ==== Stored Layers and the Map

Reviewer: ```suggestion
==== Stored layers and the map
```

> If a Polygon is clicked when loaded on the map, Geo Filter(s) will be created

Reviewer: This line seems to be out of place... perhaps this part is unfinished?

Author: Yes, there is a bigger section on geoFilters that should be added. Will make a note to get it done and when it's there, it can be referenced in a few separate places.

=== Features Omitted Due to Request Limits

> If there are more features present within the current map canvas extent than were retrieved with the set Elasticsearch query size response limit, a warning icon will appear beside the layer. The warning icon has a tooltip indicating this and the set current limit

Reviewer: ```suggestion
If there are more features in the map canvas than are allowed by the Elasticsearch query size response limit, a warning icon appears beside the layer.
```

> === More Detailed Stored Layer Configurations Example

Reviewer: ```suggestion
=== Detailed example of a stored layer configuration
```

> The icon, color, size or popupFields can be specified in the properties object of an ES document and this is the first precedence. For example, the church icon below will be a `red heart` on the map

Reviewer: ```suggestion
The `icon`, `color`, `size`, or `popupFields` can be specified in the properties object of an Elasticsearch document. For example, by using the following code, the church appears as a red heart on the map:
```

> 

Reviewer: ```suggestion
[source,json]
----
```

 {
   "spatial_path": "pois/states/churches",
   "geometry": {
     "type": "point",
     "coordinates": [ [ 100.0, 0.0 ] ]
   },
   "properties": {  
     "icon": "heart",
     "color": "red",
     "popupfields": ['Label','denomination']
   },
   "colorfield": "green"
 } 

> 

Reviewer: ```suggestion
----
```

> .Example of document with some configuration fields (example 1)

Reviewer: ```suggestion
.Example 1: A document with some configuration fields
```

> 

Reviewer: ```suggestion
[source,json]
----
```

 [
   {
     "spatial_path": "pois/states/churches",
     "icon": "cross",
     "color": ["colorfield"],
     "popupfields" : ['Label','denomination']
   },
   {
     "spatial_path": "pois/states",
     "icon": "map",
     "popupfields" : ['Label','statecode’],
     "maxZoom": 15 
   },
   {
     "color" : "default_color",
     "icon": "default_icon",
     "popupfield" : [],
     "minZoom": 0,
     "maxZoom": 18  
   }
 ]

> 

Reviewer: ```suggestion
----
```

> .Example of the Stored Layer Configuration (example 2)

Reviewer: ```suggestion
.Example 2: The stored layer configuration
```

> * Values for icon, size, popupFields and color properties on the map are taken from

Reviewer: ```suggestion
The values for the `icon`, `size`, `popupFields`, and `color` properties on the map are taken in the following order:
```

> ** Feature itself 

Reviewer: ```suggestion
. From the feature itself.
```

> ** Then from the the most specific spatial_path in the configuration

Reviewer: ```suggestion
. Then, from the the most specific `spatial_path` parameter in the configuration.
```

> ** Then from the next most relevant spatial_path in the configuration

Reviewer: ```suggestion
. Then, from the next most relevant `spatial_path` parameter in the configuration.
```

> ** Until we reach the root of the configuration where a default will be assigned

Reviewer: ```suggestion
. Finally, the root of the configuration is reached and a default value is assigned.
```

> * Values for maxZoom, minZoom properties on the map are taken from

Reviewer: ```suggestion
The values for the `maxZoom` and `minZoom` properties on the map are taken in the following order:
```

> ** Begins at the most specific spatial_path in the configuration

Reviewer: ```suggestion
. From the most specific `spatial_path` parameter in the configuration.
```

> ** Then from the next spatial_path in the configuration

Reviewer: ```suggestion
. Then, from the next `spatial_path` parameter in the configuration.
```

> ** Until we reach the root of the configuration where a default will be assigned

Reviewer: ```suggestion
. Finally, the root of the configuration is reached and a default value is assigned.
```

> * Values for Icon and color for layer on the <<Layer Control>>

Reviewer: ```suggestion
The values for `icon` and `color` for the layer in the <<Layer Control>> are taken in the same order as the `maxZoom` and `minZoom` properties.
```

> ** Same as for maxZoom, minZoom properties

Reviewer: ```suggestion
```

==== Configurations for example 1
===== For the example document above, the properties on the map will be:
* *color*: `red` - taken from the feature document
* *icon*: `heart`  - taken from the feature document
* *size*: `m` - this is the underlying default
* *popupField*: `["label", “denomination”]` - taken from the feature document
* *maxZoom*: `15` - taken from the configuration with the “pois/states” spatial_path
* *minZoom*: `0` - taken from the default configuration (the configuration without spatial_path attribute)

===== The properties on the <<Layer Control>> for example 1 will be:

> * *color*: `green` - field type, taken from the configuration and retrived from a document with the “pois/states/churches” spatial_path

Reviewer: ```suggestion
* *color*: `green` - field type, taken from the configuration and retrieved from a document with the “pois/states/churches” `spatial_path`
```
